File: 2021/12/19/main.go
package main

import "strconv"

func decodeString(s string) string {
	r, _ := dfs(0, s)
	return r
}

func dfs(i int, s string) (string, int) {
	c, r, t := "", "", ""
	for i < len(s) {
		if s[i] == '[' {
			t, i = dfs(i+1, s)
			a, _ := strconv.Atoi(c)
			r += times(a, t)
			c = ""
		} else if s[i] == ']' {
			return r, i
		} else if '0' <= s[i] && s[i] <= '9' {
			c += string(s[i])
		} else {
			r += string(s[i])
		}
		i++
	}
	return r, i
}

func times(n int, t string) string {
	r := ""
	for i := 0; i < n; i++ {
		r += t
	}
	return r
}

File: 2021/12/19/main.py
class Solution:
    def decodeString(self, s: str) -> str:
        def dfs(i):
            c = r = ''
            while i < len(s):
                if s[i] == '[':
                    t, i = dfs(i+1)
                    r += int(c) * t
                    c = ''
                elif s[i] == ']':
                    return r, i
                elif '0' <= s[i] <= '9':
                    c += s[i]
                else:
                    r += s[i]
                i += 1
            return r
        return dfs(0)
